File: src/Games/tank.ts
import { Container, Sprite, Texture } from 'pixi.js';
import Bullet from './bullet';
import { WallSegment } from './mapGenerator';

/**
 * 计算点到线段的最短距离。
 * 这是 2D 几何中的经典算法，用于坦克与墙壁的碰撞检测。
 * 为什么不直接用 Pixi.js 的 hitTest：因为墙壁是数学线段而非 PIXI.DisplayObject，
 * 且我们需要精确的"点到线段"距离而非矩形包围盒碰撞。
 *
 * @param px - 点的 X 坐标（坦克中心）
 * @param py - 点的 Y 坐标（坦克中心）
 * @param x1 - 线段起点 X
 * @param y1 - 线段起点 Y
 * @param x2 - 线段终点 X
 * @param y2 - 线段终点 Y
 * @returns 点到线段的最短距离（像素）
 */
function pointToSegmentDistance(
  px: number,
  py: number,
  x1: number,
  y1: number,
  x2: number,
  y2: number
) {
  const l2 = (x2 - x1) ** 2 + (y2 - y1) ** 2;
  if (l2 === 0) return Math.sqrt((px - x1) ** 2 + (py - y1) ** 2);
  let t = ((px - x1) * (x2 - x1) + (py - y1) * (y2 - y1)) / l2;
  t = Math.max(0, Math.min(1, t));
  return Math.sqrt(
    (px - (x1 + t * (x2 - x1))) ** 2 + (py - (y1 + t * (y2 - y1))) ** 2
  );
}

/**
 * 坦克游戏实体类，继承自 PIXI.Container。
 * Pixi.js 中"一个游戏对象 = 一个 Container"是社区标准做法，
 * 可将 Sprite、Graphics 等子对象统一管理，也便于整体变换（位移、旋转、缩放）。
 * 本类与后端 `server/src/game/tank.rs` 中的 `Tank` 结构体保持逻辑一致，
 * 确保客户端预测与权威服务器判定的行为对齐。
 */
class Tank extends Container {
  /** 炮塔精灵。使用独立 Sprite 而非合并纹理：炮塔需单独旋转（瞄准方向），车身保持固定朝向。 */
  gun: Sprite;

  /** 车身精灵。anchor 设为 (0.5, 0.5) 使旋转中心在几何中心，避免"绕角旋转"的违和感。 */
  body: Sprite;

  /** 死亡状态标记。false 表示存活，true 表示已被击中。隐藏可见性而非销毁对象：便于复活机制扩展。 */
  isDead: boolean = false;

  /** 移动速度（像素/帧）。当前固定为 3，与后端 `SPEED` 常量保持一致，确保权威服务器判定准确。 */
  speed: number = 3;

  /** 弹夹容量上限。10 发是经典坦克大战的舒适数值：足够压制，又需适时松手换弹。 */
  bulletLimit: number = 10;

  /** 移动锁定标记。当前未使用，预留用于"被击中后短暂僵直"或"道具冻结"等效果。 */
  isLocked: boolean = false;

  /** 当前剩余子弹数。长按空格持续消耗，松手瞬间重置为 `bulletLimit`。 */
  shotsRemaining: number = 10;

  /** 该坦克发出的所有存活子弹。本地模式下用于碰撞检测；多人模式下仅供本地预测显示。 */
  bullets: Bullet[] = [];

  /** 上次发射时间戳（performance.now()）。控制射速，与 `fireInterval` 配合实现冷却。 */
  lastFireTime: number = 0;

  /** 发射冷却间隔（毫秒）。200ms = 5发/秒，与后端 `FIRE_INTERVAL` 一致。 */
  fireInterval: number = 200;

  /** 空格键是否被长按。true 时 `canFire` 返回 true（若冷却完毕），实现按住连发。 */
  isSpaceHeld: boolean = false;

  /** 碰撞半径（像素）。15px 对应车身直径 30px，圆形碰撞盒计算简单且玩家可预测。 */
  radius: number = 15;

  /** 子弹失效策略。'time' = 10秒后自动消失（当前使用），'bounces' = 反弹5次后消失。 */
  bulletMode: 'bounces' | 'time' = 'time';

  /**
   * 创建坦克实例。
   *
   * @param gunTexture - 炮塔纹理，由 PIXI.Assets.load 异步加载
   * @param bodyTexture - 车身纹理
   * @param x - 初始 X 坐标，通常取地图角落（如 75）避免出生撞墙
   * @param y - 初始 Y 坐标
   *
   * 为什么用 Texture 而非直接加载路径：Pixi.js v8 要求先通过 Assets.load 加载资源，
   * 获得 Texture 后再创建 Sprite，这是 v8 的资源管理最佳实践。
   * 为什么 width/height 设为 30：对应 radius 15，直径 30px，视觉上比例协调。
   */
  constructor(gunTexture: Texture, bodyTexture: Texture, x: number, y: number) {
    super();
    this.body = new Sprite(bodyTexture);
    this.body.anchor.set(0.5);
    this.body.width = 30; // 对应 radius: 15，直径30
    this.body.height = 30;

    this.gun = new Sprite(gunTexture);
    this.gun.anchor.set(0.5, 0.5); // SVG viewBox 居中，用0.5,0.5
    this.gun.width = 30; // 和车身等宽，视觉上比例合适
    this.gun.height = 30;

    // 层级：先body再gun，gun在上层
    this.addChild(this.body, this.gun);
    this.x = x;
    this.y = y;
  }

  /**
   * 向前移动。计算当前 rotation 方向的位移向量，调用 tryMove 检测碰撞。
   * 为什么不直接修改 x/y：需要处理"碰到墙就沿墙滑动"的经典游戏体验。
   * 与后端 `server/src/game/tank.rs` 的 `move_forward` 方法逻辑一致。
   *
   * @param walls - 当前地图的所有墙壁线段
   */
  moveForward(walls: WallSegment[]) {
    const dx = Math.cos(this.rotation) * this.speed;
    const dy = Math.sin(this.rotation) * this.speed;
    this.tryMove(dx, dy, walls);
  }

  /**
   * 向后移动。与 moveForward 方向相反，使用负的速度向量。
   * 后退速度不减速：简化设计，经典坦克游戏通常前后同速。
   *
   * @param walls - 当前地图的所有墙壁线段
   */
  moveBackward(walls: WallSegment[]) {
    const dx = -Math.cos(this.rotation) * this.speed;
    const dy = -Math.sin(this.rotation) * this.speed;
    this.tryMove(dx, dy, walls);
  }

  /**
   * 尝试移动，带滑动碰撞（Sliding Collision）。
   * 这是 2D 俯视射击游戏的标配物理：
   * 1. 优先尝试完整位移（dx + dy）
   * 2. 若撞墙，尝试仅 X 轴位移
   * 3. 若仍撞墙，尝试仅 Y 轴位移
   * 4. 两轴都失败则停止
   *
   * 这种"分解移动"让玩家沿墙滑过，避免完全卡死的挫败感。
   * 与后端 `server/src/game/tank.rs` 的 `try_move` 方法逻辑一致。
   *
   * @param dx - X 轴位移量
   * @param dy - Y 轴位移量
   * @param walls - 当前地图的所有墙壁线段
   */
  private tryMove(dx: number, dy: number, walls: WallSegment[]) {
    // 优先尝试完整移动
    if (!this.checkCollision(this.x + dx, this.y + dy, walls)) {
      this.x += dx;
      this.y += dy;
      return;
    }

    // 完整移动失败，尝试只移动 X 轴
    if (!this.checkCollision(this.x + dx, this.y, walls)) {
      this.x += dx;
      return;
    }

    // 尝试只移动 Y 轴
    if (!this.checkCollision(this.x, this.y + dy, walls)) {
      this.y += dy;
    }

    // 两轴都失败则完全停止
  }

  /**
   * 向左旋转。每次减 0.05 弧度（约 2.86°）。
   * 固定步长旋转是经典设计：简单直接，玩家容易建立肌肉记忆。
   * 与后端 `server/src/game/tank.rs` 的 `turn_left` 方法逻辑一致。
   */
  turnLeft() {
    this.rotation -= 0.05;
  }

  /**
   * 向右旋转。每次加 0.05 弧度（约 2.86°）。
   * 与后端 `server/src/game/tank.rs` 的 `turn_right` 方法逻辑一致。
   */
  turnRight() {
    this.rotation += 0.05;
  }

  /**
   * 判断当前是否可以发射。包含四层判定：
   * 1. 是否死亡
   * 2. 弹夹是否为空
   * 3. 空格键是否被按住
   * 4. 冷却是否结束
   *
   * 返回 boolean 而非抛出异常：调用处通常是简单条件判断，bool 最简洁。
   * 与后端 `server/src/game/tank.rs` 的 `can_fire` 方法逻辑一致。
   *
   * @param currentTime - 当前时间戳（performance.now()）
   * @returns true 表示可以发射
   */
  canFire(currentTime: number): boolean {
    if (this.isDead) return false;
    if (this.shotsRemaining <= 0) return false;
    if (!this.isSpaceHeld) return false;

    if (currentTime - this.lastFireTime < this.fireInterval) return false;

    return true;
  }

  /**
   * 发射子弹。创建 Bullet 实例并加入场景和 bullets 数组。
   * 为什么不返回 Bullet：直接加入场景更简洁，Pixi.js 中"创建即显示"是常见模式。
   * 与后端 `server/src/game/tank.rs` 的 `fire` 方法逻辑一致，
   * 但后端返回 Option<Bullet> 以便解耦测试，前端直接显示。
   *
   * @param stage - Pixi.js 场景容器，子弹将被添加到此容器中
   * @param currentTime - 当前时间戳（performance.now()）
   */
  fire(stage: Container, currentTime: number) {
    if (!this.canFire(currentTime)) return;

    const bullet = new Bullet(this.x, this.y, this.rotation, this.bulletMode);
    stage.addChild(bullet);
    this.bullets.push(bullet);
    this.shotsRemaining--;
    this.lastFireTime = currentTime;
  }

  /**
   * 空格键按下事件。设置 isSpaceHeld = true，启动连发状态。
   * 为什么有 onSpaceDown/onSpaceUp 两个方法：分离按键按下和松开逻辑，
   * 是游戏输入处理的标准做法（Input Event Pattern），便于扩展"短按/长按"区分。
   * 与后端 `server/src/game/tank.rs` 的 `on_fire_down` 方法对应。
   */
  onSpaceDown() {
    this.isSpaceHeld = true;
  }

  /**
   * 空格键松开事件。重置 isSpaceHeld 并补满弹夹。
   * "松手换弹"机制给玩家策略选择：持续压制（长按）or 节奏点射（短按）。
   * 与后端 `server/src/game/tank.rs` 的 `on_fire_up` 方法对应。
   */
  onSpaceUp() {
    this.isSpaceHeld = false;
    this.shotsRemaining = this.bulletLimit;
  }

  /**
   * 更新子弹状态并清理失效子弹。本地模式下每帧调用。
   * 倒序遍历（i--）：删除数组元素时不影响未遍历的索引，是数组删除的标准技巧。
   * 与后端 `server/src/game/tank.rs` 的 `update_bullets` 方法逻辑一致。
   *
   * @param walls - 当前地图的所有墙壁线段
   */
  update(walls: WallSegment[]) {
    for (let i = this.bullets.length - 1; i >= 0; i--) {
      const b = this.bullets[i];
      b.update(walls);

      // 子弹失活 → 从场景和数组中移除
      if (!b.active) {
        b.parent?.removeChild(b);
        b.destroy();
        this.bullets.splice(i, 1);
      }
    }
  }

  /**
   * 检测给定坐标是否与墙壁碰撞。
   * 使用"点到线段距离"算法：计算坦克中心到每面墙的最短距离，小于 radius 则撞墙。
   * 为什么不用 AABB（轴对齐包围盒）：地图墙壁是任意线段，AABB 只适用于矩形碰撞。
   * 与后端 `server/src/game/collision.rs` 的 `check_collision` 函数逻辑一致。
   *
   * @param nx - 待检测的 X 坐标
   * @param ny - 待检测的 Y 坐标
   * @param walls - 当前地图的所有墙壁线段
   * @returns true 表示发生碰撞
   */
  checkCollision(nx: number, ny: number, walls: WallSegment[]): boolean {
    for (const wall of walls) {
      // 计算坦克中心点到墙线段的距离
      const dist = pointToSegmentDistance(
        nx,
        ny,
        wall.x1,
        wall.y1,
        wall.x2,
        wall.y2
      );
      // 如果距离小于坦克半径，则视为撞墙
      if (dist < this.radius) return true;
    }
    return false;
  }

  /**
   * 更新子弹 + 碰撞检测的合并方法。本地模式主循环调用。
   * 与 update() 的区别：update() 只做子弹更新，此方法额外包含碰撞检测逻辑。
   * 保留两个版本：update() 便于测试，updateWithCollision() 供主循环使用。
   * 与后端 `server/src/game/state.rs` 的 `process_tick` 中的逻辑对应。
   *
   * @param walls - 当前地图的所有墙壁线段
   */
  updateWithCollision(walls: WallSegment[]) {
    this.bullets.forEach((b) => b.update(walls));
    for (let i = this.bullets.length - 1; i >= 0; i--) {
      const b = this.bullets[i];
      b.update(walls);

      // 子弹失活 → 从场景和数组中移除
      if (!b.active) {
        b.parent?.removeChild(b);
        b.destroy();
        this.bullets.splice(i, 1);
      }
    }
  }

  /**
   * 坦克死亡。设置 isDead 并隐藏可见性。
   * 不销毁对象：保留 Tank 实例便于复活机制，且避免 PIXI 对象频繁创建/销毁的性能开销。
   * 与后端 `server/src/game/tank.rs` 的 `die` 方法对应。
   */
  die() {
    this.isDead = true;
    this.visible = false;
  }
}

export default Tank;

File: server/src/game/tank.rs
// Tank Game — 坦克大战
// Copyright (C) 2026 Una
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//! Tank logic (mirrored from frontend)
//! 本模块与前端 `src/Games/tank.ts` 保持逻辑一致，确保权威服务器判定与客户端预测对齐。

use crate::game::collision::check_collision;
use crate::game::map::WallSegment;

/// 基础移动速度（像素/ tick）。3.0 是经典坦克大战的舒适速度：
/// 既不会太快导致操控困难，也不会太慢让玩家感到拖沓。
const SPEED: f64 = 3.0;

/// 每次旋转的弧度增量。0.05 弧度 ≈ 2.86°，属于固定转角速度设计。
/// 为什么不用时序无关的旋转（如 delta_time * speed）：简单直接，经典坦克游戏通常采用固定步长。
const ROTATION_SPEED: f64 = 0.05;

/// 坦克碰撞半径（像素）。15.0 对应车身直径 30px，与前端 `radius: 15` 一致。
/// 圆形碰撞盒是俯视射击游戏的行业惯例，计算简单且玩家可预测。
const RADIUS: f64 = 15.0;

/// 发射冷却间隔（毫秒）。200ms = 5发/秒，与前端 `fireInterval: 200` 一致。
/// 这是经典坦克大战的舒适射速：足够压制，又需节奏控制。
const FIRE_INTERVAL: u64 = 200;

/// 弹夹容量上限。10 发与前端 `bulletLimit: 10` 一致。
const BULLET_LIMIT: i32 = 10;

/// 子弹生成位置偏移量（像素）。30.0 确保子弹不在坦克内部生成，避免自伤。
/// 与前端 `this.x + Math.cos(angle) * 30` 对应。
const BULLET_SPAWN_OFFSET: f64 = 30.0;

/**
 * 服务端坦克实体。
 * 作为权威服务器（Authoritative Server）的核心对象，所有移动和射击判定以本结构状态为准。
 * 与前端 `src/Games/tank.ts` 中的 `Tank` 类保持逻辑一致，字段命名和语义尽量对齐，
 * 便于前后端联调时快速定位差异。
 */
pub struct Tank {
    /// X 轴坐标。使用 f64 而非 i32：游戏物理计算涉及大量三角函数（cos/sin），
    /// 浮点数可直接参与运算，避免频繁类型转换。Rust 游戏开发社区普遍使用 f64。
    pub x: f64,

    /// Y 轴坐标。与 x 保持同类型，确保坐标运算一致性。
    pub y: f64,

    /// 坦克朝向，单位为弧度（radians）。0 表示正右方，递增为顺时针旋转。
    /// 使用弧度而非角度：Rust 标准库三角函数只接受弧度，省去每帧 deg→rad 转换开销。
    pub rotation: f64,

    /// 移动速度（像素/ tick）。当前固定为 SPEED，未来可扩展为可变速度（如加速道具）。
    /// 提取为字段而非硬编码：便于运行时修改，符合"数据驱动"设计习惯。
    pub speed: f64,

    /// 死亡标记。使用 bool 而非 Option<Tank>：死亡是游戏中的常见状态，
    /// bool 占用 1 字节且判断分支简单，比 Option 更直观高效。
    pub is_dead: bool,

    /// 碰撞半径（像素）。值为 15.0，对应车身直径 30px。
    /// 与前端 `radius: number = 15` 一致，确保碰撞判定跨端对齐。
    pub radius: f64,

    /// 当前存活的子弹列表。使用 Vec 而非对象池：4 人 × 每坦克 10 弹 = 40 个对象，
    /// 现代 allocator 可轻松处理。若未来扩展为 32 人大房间，可考虑换为 slab 结构。
    pub bullets: Vec<super::Bullet>,

    /// 剩余可发射子弹数。使用 i32 而非 u32：防止无符号减法下溢 panic，
    /// 尤其在 debug 模式下 Rust 会检查整数溢出。游戏开发社区习惯用有符号整数做计数。
    pub shots_remaining: i32,

    /// 上次发射时间戳（毫秒，Unix epoch）。用于控制射速，与 `fire_interval` 配合。
    /// 为什么不使用 std::time::Instant：Instant 不可序列化，而 timestamp 便于网络同步和日志记录。
    pub last_fire_time: u64,

    /// 发射冷却间隔（毫秒）。与常量 FIRE_INTERVAL 一致，提取为字段便于运行时修改（如道具影响）。
    pub fire_interval: u64,

    /// 开火键是否被按住。长按空格可持续发射，松开时重置弹夹（`shots_remaining`）。
    /// 这是街机射击游戏的经典机制：按住连发、松手换弹。
    pub is_fire_held: bool,

    /// 子弹失效模式。Time 表示 10 秒后自动消失，Bounces 表示反弹 5 次后消失。
    /// 提取为枚举而非 bool：未来可轻松扩展新模式（如穿透、追踪等）。
    pub bullet_mode: super::BulletMode,
}

impl Tank {
    /**
     * 创建新坦克实例。
     *
     * @param x - 初始 X 坐标，通常取地图角落（如 75.0），给玩家留出足够空间避免出生即撞墙
     * @param y - 初始 Y 坐标
     *
     * 为什么 start_x/start_y 不用随机：固定出生点让玩家有策略预期，
     * 且便于平衡（角落出生相对安全）。
     */
    pub fn new(x: f64, y: f64) -> Self {
        Self {
            x,
            y,
            rotation: 0.0,
            speed: SPEED,
            is_dead: false,
            radius: RADIUS,
            bullets: Vec::new(),
            shots_remaining: BULLET_LIMIT,
            last_fire_time: 0,
            fire_interval: FIRE_INTERVAL,
            is_fire_held: false,
            bullet_mode: super::BulletMode::Time,
        }
    }

    /// 获取 X 坐标。提供 getter 便于未来添加偏移计算（如后坐力位移）。
    pub fn x(&self) -> f64 {
        self.x
    }

    /// 获取 Y 坐标。
    pub fn y(&self) -> f64 {
        self.y
    }

    /// 获取朝向（弧度）。
    pub fn rotation(&self) -> f64 {
        self.rotation
    }

    /// 获取死亡状态。
    pub fn is_dead(&self) -> bool {
        self.is_dead
    }

    /// 获取剩余子弹数。
    pub fn shots_remaining(&self) -> i32 {
        self.shots_remaining
    }

    /**
     * 向前移动。计算当前朝向的位移向量，调用 `try_move` 进行碰撞检测。
     * 为什么不直接修改 x/y：需要处理"碰到墙就沿墙滑动"的经典游戏体验。
     * 与前端 `src/Games/tank.ts` 的 `moveForward` 方法逻辑一致。
     *
     * @param walls - 当前地图的所有墙壁线段
     */
    pub fn move_forward(&mut self, walls: &[WallSegment]) {
        let dx = self.rotation.cos() * self.speed;
        let dy = self.rotation.sin() * self.speed;
        self.try_move(dx, dy, walls);
    }

    /**
     * 向后移动。与 move_forward 方向相反，使用负的速度向量。
     * 后退速度不减速：简化设计，经典坦克游戏通常前后同速。
     * 与前端 `src/Games/tank.ts` 的 `moveBackward` 方法逻辑一致。
     *
     * @param walls - 当前地图的所有墙壁线段
     */
    pub fn move_backward(&mut self, walls: &[WallSegment]) {
        let dx = -self.rotation.cos() * self.speed;
        let dy = -self.rotation.sin() * self.speed;
        self.try_move(dx, dy, walls);
    }

    /**
     * 尝试移动，带滑动碰撞（Sliding Collision）。
     * 这是 2D 俯视游戏的标准做法：优先尝试完整移动，若撞墙则分别尝试 X/Y 单轴移动，
     * 给玩家"沿墙滑过"的流畅感，而非完全卡死。
     * 与前端 `src/Games/tank.ts` 的 `tryMove` 方法逻辑一致。
     *
     * @param dx - X 轴位移量
     * @param dy - Y 轴位移量
     * @param walls - 当前地图的所有墙壁线段
     */
    fn try_move(&mut self, dx: f64, dy: f64, walls: &[WallSegment]) {
        if !check_collision(self.x + dx, self.y + dy, self.radius, walls) {
            self.x += dx;
            self.y += dy;
            return;
        }

        if !check_collision(self.x + dx, self.y, self.radius, walls) {
            self.x += dx;
            return;
        }

        if !check_collision(self.x, self.y + dy, self.radius, walls) {
            self.y += dy;
        }
    }

    /**
     * 向左旋转。每次减 ROTATION_SPEED 弧度（约 2.86°）。
     * 固定步长旋转是经典设计：简单直接，玩家容易建立肌肉记忆。
     * 与前端 `src/Games/tank.ts` 的 `turnLeft` 方法逻辑一致。
     */
    pub fn turn_left(&mut self) {
        self.rotation -= ROTATION_SPEED;
    }

    /**
     * 向右旋转。每次加 ROTATION_SPEED 弧度。
     * 与前端 `src/Games/tank.ts` 的 `turnRight` 方法逻辑一致。
     */
    pub fn turn_right(&mut self) {
        self.rotation += ROTATION_SPEED;
    }

    /**
     * 判断当前是否可以发射。包含死亡检查、弹夹检查、按住检查、冷却检查四层过滤。
     * 返回 bool 而非 Result：调用处通常是"能发就发"的简单逻辑，bool 更简洁。
     * 与前端 `src/Games/tank.ts` 的 `canFire` 方法逻辑一致。
     *
     * @param current_time - 当前时间戳（毫秒，Unix epoch）
     * @returns true 表示可以发射
     */
    pub fn can_fire(&self, current_time: u64) -> bool {
        if self.is_dead {
            return false;
        }
        if self.shots_remaining <= 0 {
            return false;
        }
        if !self.is_fire_held {
            return false;
        }
        if current_time - self.last_fire_time < self.fire_interval {
            return false;
        }
        true
    }

    /**
     * 发射子弹。生成 Bullet 实例并递减弹夹。
     * 返回 Option<Bullet> 而非直接 push：解耦"发射判定"与"子弹创建"，便于测试和复用。
     * 与前端 `src/Games/tank.ts` 的 `fire` 方法逻辑一致，
     * 但前端直接加入场景（Pixi.js 习惯），后端返回 Option 以便 GameState 统一管理。
     *
     * @param id - 子弹唯一 ID，由 GameState 分配
     * @param current_time - 当前时间戳（毫秒，Unix epoch）
     * @returns Some(Bullet) 表示发射成功，None 表示条件不满足
     */
    pub fn fire(&mut self, id: u32, current_time: u64) -> Option<super::Bullet> {
        if !self.can_fire(current_time) {
            return None;
        }

        let bullet = super::Bullet::new(
            id,
            self.x + self.rotation.cos() * BULLET_SPAWN_OFFSET,
            self.y + self.rotation.sin() * BULLET_SPAWN_OFFSET,
            self.rotation,
            self.bullet_mode,
        );

        self.shots_remaining -= 1;
        self.last_fire_time = current_time;

        Some(bullet)
    }

    /**
     * 更新该坦克的所有子弹状态。每 tick 调用一次，清理已失效子弹。
     * 为什么不放在 GameState 里统一更新：保持 Tank 的自包含性，便于单机测试。
     * 与前端 `src/Games/tank.ts` 的 `update` 方法逻辑一致。
     *
     * @param walls - 当前地图的所有墙壁线段
     * @param current_time - 当前时间戳（毫秒，Unix epoch）
     */
    pub fn update_bullets(&mut self, walls: &[WallSegment], current_time: u64) {
        for bullet in &mut self.bullets {
            bullet.update(walls, current_time);
        }

        self.bullets.retain(|b| b.active);
    }

    /**
     * 标记开火键按下。同时设置 `is_fire_held = true`，进入连发状态。
     * 与前端 `src/Games/tank.ts` 的 `onSpaceDown` 方法对应。
     */
    pub fn on_fire_down(&mut self) {
        self.is_fire_held = true;
    }

    /**
     * 标记开火键松开。重置 `is_fire_held` 并补满弹夹。
     * 这是"按住连发、松手换弹"机制的核心：给玩家策略选择——持续压制 or 节奏点射。
     * 与前端 `src/Games/tank.ts` 的 `onSpaceUp` 方法对应。
     */
    pub fn on_fire_up(&mut self) {
        self.is_fire_held = false;
        self.shots_remaining = BULLET_LIMIT;
    }

    /**
     * 坦克死亡。设置 is_dead 为 true。
     * 不销毁对象：保留 Tank 实例便于复活机制，且避免频繁重建的性能开销。
     * 与前端 `src/Games/tank.ts` 的 `die` 方法对应。
     */
    pub fn die(&mut self) {
        self.is_dead = true;
    }
}
